scripts/tfsenc_encoding.py

```diff
@@ -0,0 +1,243 @@
+from himalaya.kernel_ridge import (
+    MultipleKernelRidgeCV,
+    KernelRidgeCV,
+    ColumnKernelizer,
+    Kernelizer,
+)
+from himalaya.scoring import correlation_score, correlation_score_split
+
+
+@jit(nopython=True)
+def build_Y(brain_signal, onsets, lags, window_size):
+    """[summary]
+
+    Args:
+        onsets ([type]): [description]
+        brain_signal ([type]): [description]
+        lags ([type]): [description]
+        window_size ([type]): [description]
+
+    Returns:
+        [type]: [description]
+    """
+    half_window = round((window_size / 1000) * 512 / 2)  # convert to signal fs
+    Y1 = np.zeros((len(onsets), len(lags)))
+
+    for lag in prange(len(lags)):
+        lag_amount = int(lags[lag] / 1000 * 512)  # convert fo signal fs
+        index_onsets = np.round_(onsets, 0, onsets) + lag_amount  # lag onsets
+        starts = index_onsets - half_window - 1  # lag window onset
+        stops = index_onsets + half_window  # lag window offset
+        for i, (start, stop) in enumerate(zip(starts, stops)):
+            Y1[i, lag] = np.mean(
+                brain_signal[start:stop].reshape(-1)
+            )  # average lag window
+
+    return Y1
+
+
+def get_groupkfolds(datum, X, Y, fold_num=10):
+    fold_cat = np.zeros(datum.shape[0])
+    grpkfold = GroupKFold(n_splits=fold_num)
+    folds = [t[1] for t in grpkfold.split(X, Y, groups=datum["conversation_id"])]
+
+    for i in range(0, len(folds)):
+        for row in folds[i]:
+            fold_cat[row] = i  # turns into fold category
+
+    fold_cat_comp = fold_cat[datum.speaker != "Speaker1"]
+    fold_cat_prod = fold_cat[datum.speaker == "Speaker1"]
+
+    return (fold_cat_comp, fold_cat_prod)
+
+
+def get_kfolds(X, fold_num=10):
+    print("Using kfolds")
+    skf = KFold(n_splits=fold_num, shuffle=False)
+    folds = [t[1] for t in skf.split(np.arange(X.shape[0]))]
+    fold_cat = np.zeros(X.shape[0])
+    for i in range(0, len(folds)):
+        for row in folds[i]:
+            fold_cat[row] = i  # turns into fold category
+    return fold_cat
+
+
+def encoding_setup(args, elec_name, elec_datum, elec_signal):
+
+    # Build x and y matrices
+    X = np.stack(elec_datum.embeddings).astype("float64")
+    Y = build_Y(
+        elec_signal.reshape(-1, 1),
+        elec_datum.adjusted_onset.values,
+        np.array(args.lags),
+        args.window_size,
+    )
+    X = X.astype("float32")
+    Y = Y.astype("float32")
+
+    # Split into production and comprehension
+    prod_X = X[elec_datum.speaker == "Speaker1", :]
+    comp_X = X[elec_datum.speaker != "Speaker1", :]
+    prod_Y = Y[elec_datum.speaker == "Speaker1", :]
+    comp_Y = Y[elec_datum.speaker != "Speaker1", :]
+
+    # get folds
+    if args.project_id == "podcast":  # podcast
+        fold_cat_comp = get_kfolds(comp_X, args.cv_fold_num)
+        fold_cat_prod = []
+    elif len(args.conv_ids) < args.cv_fold_num * 2:  # small num of convos, also 798
```

Comment:
why do we use kfold specifically when there's fewer than 2 * cv_fold_num conversations? in principle we could use group kfolds if conv_ids == cv_fold_num, right?

Reply:
because the folds come from all the conversations pooled and also the number of words is different in different conversations.

Reply:
We probably want some flexibility. If there are 10 folds and we have 10 conversations, it's basically one convo per fold and not too flexible. If we have at least 20 conversations and they are assigned to 10 folds based on groupkfolds, then there is some flexibility. If you want, we can also just ask for conv_ids >= cv_fold_num

Reply:
We could do groupkfolds by default, and let the user specify via an arg if they want to use kfolds